### ISSUE TITLE
After null safety migaration compiler can't infer types that it could before

### ISSUE THREAD

**Issue by escamoteur** [dart-lang/sdk#44825]
I'm currently migrating my get_it_mixin package. After applying the changes I got a lot of errors in my tests that the compiler couldn't infer several types anymore. After making some of them nullable again most issues went away, but with one function that gets a handler passed the inference fails although it worked witout problems before:

<img width="869" alt="grafik" src="https://user-images.githubusercontent.com/16635729/106490535-3a4ecb80-64b6-11eb-9620-547d419ade22.png">

```Dart
  void registerStreamHandler<T extends Object, R>(
    Stream<R> Function(T) select,
    void Function(BuildContext? context, AsyncSnapshot<R> newValue,
            void Function() cancel)
        handler, {
    R? initialValue,
    String? instanceName,
  }) =>
```

Overall I feel migrating Function types quite difficult.

You can have a look at the current commit of the repository and comare it to the non safe master branch.

https://github.com/escamoteur/get_it_mixin/commit/cdc6627f392cd0e9239d21af3e0b0ebf6527a4d6

I'm running on latest master:

```
[✓] Flutter (Channel master, 1.26.0-18.0.pre.92, on macOS 11.1 20C69 darwin-arm, locale de-DE)
    • Flutter version 1.26.0-18.0.pre.92 at /Users/escmoteur/Entwicklung/flutter
    • Framework revision d968e33b81 (5 hours ago), 2021-02-01 07:21:22 -0500
    • Engine revision ce14c8a841
    • Dart version 2.12.0 (build 2.12.0-276.0.dev)
```

I really would like to have it working as before but I have no idea how to do it.

**Comment by escamoteur** [comment]
@leafpetersen If someone could give me some help here it would be awesome

**Comment by eernstg** [comment]
You might well have relied on implicit downcasts in the pre-null-safety version, and those locations will now require an explicit cast (that seems likely to be the case for `s.data`). You can add those casts, which will preserve the static analysis that you had previously (the only difference is that the source code will explicitly show that there is a dynamic type check at this location, it used to be implicit).

You could also try to improve on the typing, e.g., by changing some declared types, type parameter bounds, adding some explicit type arguments that were previously inferred, etc, until you have a program with so few downcasts that you're satisfied. But you can't tell up front how many changes are required for that, so you'd probably want to do it in specific, limited steps.

**Comment by escamoteur** [comment]
Thanks, I have a closer look. I will close this for now. If I don't get along I'll comming back

**Comment by leafpetersen** [comment]
This is an instance of [this problem](https://github.com/dart-lang/language/issues/731), unless I'm misunderstood your code.

TL;DR - Inference was never working there, but before we would default to `dynamic` for `s` and the code would just work (but with no static checking if you got it wrong, and no code completion).  Now we default to `Object?` for `s` which means the code gets statically rejected.  

See that issue for more discussion, including what we can do to improve this at some point in the future.

**Comment by escamoteur** [comment]
@leafpetersen yout are absolutely correct. I tightened the types to extending Object while working on the null safety that was the real reason.
Still it would be really awesome if we could solve the issue you linked.
